### ISSUE TITLE
Java analysis extraction fails to `getCanonicalClassname` for nested object

### ISSUE THREAD

**Issue by stuhood** [sbt/zinc#355]
While attempting to compile Java code that depends on `com.twitter.io.Buf.Composite.Impl` ([defined here](https://github.com/twitter/util/blob/44128deb7ad989be36f825612cc958599cfe3af7/util-core/src/main/scala/com/twitter/io/Buf.scala#L475)), zinc crashes with a stack like the following:
```
[error] Malformed class name
[error] java.lang.Class.getSimpleName(Class.java:1330)
[error] java.lang.Class.getCanonicalName(Class.java:1399)
[error] sbt.internal.inc.ClassToAPI$.classCanonicalName(ClassToAPI.scala:77)
[error] sbt.internal.inc.ClassToAPI$.toDefinitions(ClassToAPI.scala:80)
[error] sbt.internal.inc.ClassToAPI$.$anonfun$structure$5(ClassToAPI.scala:148)
[error] scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
[error] scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
[error] scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
[error] scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:38)
[error] scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
[error] scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
[error] scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
[error] sbt.internal.inc.ClassToAPI$.merge(ClassToAPI.scala:369)
[error] sbt.internal.inc.ClassToAPI$.structure(ClassToAPI.scala:150)
[error] sbt.internal.inc.ClassToAPI$.x$2$lzycompute$1(ClassToAPI.scala:92)
[error] sbt.internal.inc.ClassToAPI$.x$2$1(ClassToAPI.scala:92)
[error] sbt.internal.inc.ClassToAPI$.instance$lzycompute$1(ClassToAPI.scala:92)
[error] sbt.internal.inc.ClassToAPI$.instance$1(ClassToAPI.scala:92)
[error] sbt.internal.inc.ClassToAPI$.$anonfun$toDefinitions0$1(ClassToAPI.scala:99)
[error] xsbti.api.SafeLazyProxy$$anon$1.apply(SafeLazyProxy.scala:24)
[error] xsbti.api.SafeLazy$Impl.get(SafeLazy.java:55)
[error] sbt.internal.inc.ClassToAPI$.$anonfun$process$2(ClassToAPI.scala:29)
[error] scala.collection.immutable.List.foreach(List.scala:389)
[error] scala.collection.generic.TraversableForwarder.foreach(TraversableForwarder.scala:35)
[error] scala.collection.generic.TraversableForwarder.foreach$(TraversableForwarder.scala:35)
[error] scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:44)
[error] sbt.internal.inc.ClassToAPI$.process(ClassToAPI.scala:29)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.readAPI$1(AnalyzingJavaCompiler.scala:142)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$20(AnalyzingJavaCompiler.scala:160)
[error] sbt.internal.inc.classfile.Analyze$.readInheritanceDependencies$1(Analyze.scala:134)
[error] sbt.internal.inc.classfile.Analyze$.$anonfun$apply$16(Analyze.scala:150)
[error] sbt.internal.inc.classfile.Analyze$.$anonfun$apply$16$adapted(Analyze.scala:67)
[error] scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:789)
[error] scala.collection.mutable.HashMap.$anonfun$foreach$1(HashMap.scala:138)
[error] scala.collection.mutable.HashTable.foreachEntry(HashTable.scala:236)
[error] scala.collection.mutable.HashTable.foreachEntry$(HashTable.scala:229)
[error] scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:40)
[error] scala.collection.mutable.HashMap.foreach(HashMap.scala:138)
[error] scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:788)
[error] sbt.internal.inc.classfile.Analyze$.apply(Analyze.scala:67)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$19(AnalyzingJavaCompiler.scala:160)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$19$adapted(AnalyzingJavaCompiler.scala:158)
[error] scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:789)
[error] scala.collection.immutable.List.foreach(List.scala:389)
[error] scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:788)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$17(AnalyzingJavaCompiler.scala:158)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.timed(AnalyzingJavaCompiler.scala:195)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.compile(AnalyzingJavaCompiler.scala:158)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:104)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:132)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileJava$1(MixedAnalyzingCompiler.scala:89)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:115)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:304)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:304)
[error] sbt.internal.inc.Incremental$.doCompile(Incremental.scala:101)
[error] sbt.internal.inc.Incremental$.$anonfun$compile$4(Incremental.scala:82)
[error] sbt.internal.inc.IncrementalCommon.recompileClasses(IncrementalCommon.scala:118)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:64)
[error] sbt.internal.inc.Incremental$.$anonfun$compile$3(Incremental.scala:84)
[error] sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:129)
[error] sbt.internal.inc.Incremental$.compile(Incremental.scala:75)
[error] sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:70)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:308)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:266)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:157)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:236)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:67)
[error] org.pantsbuild.zinc.compiler.Main$.main(Main.scala:109)
[error] org.pantsbuild.zinc.compiler.Main.main(Main.scala)
```

~~This _may_ be a red herring, because I was working in a branch that combines both https://github.com/sbt/zinc/pull/325 and https://github.com/sbt/zinc/pull/351 (a prerelease version: [1.0.0-X19-M1](https://bintray.com/scalacenter/releases/zinc/1.0.0-X19-M1))~~ EDIT: Confirm on `X20` as well, but the location of the crash looks completely unrelated to those changes, so thought it would be worth posting now.

**Comment by stuhood** [comment]
cc @jvican

**Comment by jvican** [comment]
I'm looking into this as soon as I'm done with the API work. This seems an issue originated from the fact that Zinc is more precise now than it was in X16, and it's exploring a path that causes a [well-known behaviour in scalac](https://issues.scala-lang.org/browse/SI-5425).
  
I need to diagnose whether:
1. That exception is legit and we should not call `getCanonicalName` in certain classes; or
2. That path should never be explored by Zinc; or
3. A mix of the two.

**Comment by jvican** [comment]
I'm having a look at this this week. My strategy is to check if I can reproduce this issue with sbt 1.0.0-RC2, which will involve fixing Twitter util's build. The reason why I want to reproduce with sbt is to be certain that this is an issue that affects both Pants and sbt. If I can reproduce, I'll start having a closer look at why this is happening and what we can do to fix it.

**Comment by stuhood** [comment]
For the record: I repro this on the official X20 release.

**Comment by stuhood** [comment]
@jvican : It should not require fixing `util`'s build, afaik: if you have that library as a binary/versioned dep for some java code built with zinc, you should be able to repro.

Alternatively, as soon as I get https://github.com/pantsbuild/pants/pull/4728 landed I should be able create a repro pants build.

**Comment by jvican** [comment]
Weird @stuhood, I'm not able to reproduce... See https://github.com/sbt/zinc/pull/374. I wanted first to reproduce out of the Pants integration so that if there's an error there we can spot sooner rather than later.

**Comment by jvican** [comment]
So I haven't been able to reproduce this either in twitter-util with sbt 1.0.0-RC2. Here's the commit that makes the switch: https://github.com/jvican/util/commit/3e925e3bf940b7db5f8c8f3682b117d74dd76cd5.
  
The proof:
```
sbt:twitter-util> util-core/compile
[info] Updating {file:/data/rw/code/scala/twitter-util/}util-function...

[info] :: loading settings :: url = jar:file:/home/jvican/.sbt/boot/scala-2.12.2/org.scala-sbt/sbt/1.0.0-RC2/ivy-2.3.0-sbt-a3314352b638afbf0dca19f127e8263ed6f898bd.jar!/org/apache/ivy/core/settings/ivysettings.xml

[info] :: loading settings :: url = jar:file:/home/jvican/.sbt/boot/scala-2.12.2/org.scala-sbt/sbt/1.0.0-RC2/ivy-2.3.0-sbt-a3314352b638afbf0dca19f127e8263ed6f898bd.jar!/org/apache/ivy/core/settings/ivysettings.xml

[info] :: loading settings :: url = jar:file:/home/jvican/.sbt/boot/scala-2.12.2/org.scala-sbt/sbt/1.0.0-RC2/ivy-2.3.0-sbt-a3314352b638afbf0dca19f127e8263ed6f898bd.jar!/org/apache/ivy/core/settings/ivysettings.xml

[info] :: loading settings :: url = jar:file:/home/jvican/.sbt/boot/scala-2.12.2/org.scala-sbt/sbt/1.0.0-RC2/ivy-2.3.0-sbt-a3314352b638afbf0dca19f127e8263ed6f898bd.jar!/org/apache/ivy/core/settings/ivysettings.xml

[info] Done updating.

[info] Updating {file:/data/rw/code/scala/twitter-util/}util-core...

[info] Done updating.

[info] Compiling 5 Java sources to /data/rw/code/scala/twitter-util/util-function/target/scala-2.12/classes ...

[info] Done compiling.

[info] Compiling 87 Scala sources and 15 Java sources to /data/rw/code/scala/twitter-util/util-core/target/scala-2.12/classes ...

[info] Non-compiled module 'compiler-bridge_2.12' for Scala 2.12.1. Compiling...

[info]   Compilation completed in 9.691s.

[warn] there was one deprecation warning (since 2.12.0)

[warn] there were 13 deprecation warnings (since 2016-11-21)

[warn] there were 14 deprecation warnings in total; re-run with -deprecation for details

[warn] three warnings found

[info] Done compiling.

[info] Total time: 26 s, completed Jul 20, 2017 6:00:55 PM

sbt:twitter-util> % 
```
  
This bug looks like an issue in the Pants side, but I'm not sure what could be wrong... :disappointed:.

**Comment by stuhood** [comment]
AFAICT, it requires two compilation units: #374 does not contain a java file, so probably won't hit it. Will try to get a repro today.

EDIT: Sorry, should have included more info in the original report, but: the specific module that fails to compile is `util/util-core/src/test/java`.
EDIT2: And the specific file that fails is `util/util-core/src/test/java/com/twitter/io/BufCompilationTest.java`.

**Comment by stuhood** [comment]
The following commit, (when compiled with X20) is sufficient to repro: https://github.com/twitter/pants/commit/0693b06ed957ca43590b3db9457b603f926a93d1 ... ie, one 3rdparty binary jar, and one source file. Should be easy to extract into an sbt project.

**Comment by jvican** [comment]
https://github.com/scalacenter/zinc/commit/13f6fc658cb05de71caa650057fa37d98a494adc reproduces it now. Thanks for those last pointers and the reproduction case for Pants.

**Comment by jvican** [comment]
Weird. If the compiled scala file is refactored to the bare minimum:
```scala
///  COPIED FROM TWITTER UTIL TO IDENTIFY ISSUE WITH MALFORMED NAME  //
///////////////////////////////////////////////////////////////////////

package repro

import java.nio.ReadOnlyBufferException
import java.nio.charset.{Charset, StandardCharsets => JChar}
import scala.collection.immutable.VectorBuilder

abstract class Buf { outer =>
  @throws(classOf[IllegalArgumentException])
  def write(output: Array[Byte], off: Int): Unit

  @throws(classOf[IllegalArgumentException])
  @throws(classOf[ReadOnlyBufferException])
  def write(output: java.nio.ByteBuffer): Unit

  def length: Int
  def slice(from: Int, until: Int): Buf
  def get(index: Int): Byte
  def process(from: Int, until: Int, processor: Buf.Processor): Int

  protected def unsafeByteArrayBuf: Option[Buf.ByteArray]
}

object Buf {
  abstract class Processor {
    def apply(byte: Byte): Boolean
  }

  class ByteArray(
      private[Buf] val bytes: Array[Byte],
      private[Buf] val begin: Int,
      private[Buf] val end: Int)
    extends Buf {

    def get(index: Int): Byte = ???
    def process(from: Int, until: Int, processor: Processor): Int = ???
    def write(buf: Array[Byte], off: Int): Unit = ???
    def write(buffer: java.nio.ByteBuffer): Unit = ???

    def slice(from: Int, until: Int): Buf = ???
    def length: Int = end - begin
    protected def unsafeByteArrayBuf: Option[Buf.ByteArray] = ???
  }
}
```
  
It does not crash.

**Comment by jvican** [comment]
**EDIT**: No, false positive.
 
<s>I found the source of the error at the first attempt (pure luck). It looks as if adding this class to the companion object of `Buf` makes Zinc fail:
```scala
  private class NoopBuf extends Buf {
    def write(buf: Array[Byte], off: Int): Unit =
      checkWriteArgs(buf.length, off)

    def write(buffer: java.nio.ByteBuffer): Unit = ()

    override val isEmpty = true
    def length: Int = 0
    def slice(from: Int, until: Int): Buf = {
      checkSliceArgs(from, until)
      this
    }
    protected def unsafeByteArrayBuf: Option[Buf.ByteArray] = None
    def get(index: Int): Byte =
      throw new IndexOutOfBoundsException(s"Index out of bounds: $index")
    def process(from: Int, until: Int, processor: Processor): Int = {
      checkSliceArgs(from, until)
      -1
    }
  }
```
</s>

**Comment by jvican** [comment]
I've engineered a minimal test case that shows the issue in https://github.com/sbt/zinc/pull/374/commits/adc211478bf7bd4594a0406cdd2ed0ccf52c4b12.
  
The code fails in latest Zinc because of this commit https://github.com/sbt/zinc/commit/29571f82252f496ee2936f2d6ff7cad4225fe154, that restored Java analysis after the new analysis mechanism was introduced.
  
I'm working on a fix. This is a legit compiler error, known for years: https://github.com/scala/bug/issues/2034. I am thinking on the best way to solve this issue without affecting correctness. From a cursory analysis, it seems that the current `ClassToAPI` is not built to support Java class names inside objects.

**Comment by jvican** [comment]
@stuhood I believe https://github.com/scalacenter/zinc/commit/134a8baeaa4d84cc4e5fba1d47284fd5c5530152 fixes this thing. I'm asking for feedback on my fix. I think it's the only possible one. If we really want to address this issue, we must fix it in upstream Scala, and that won't happen before 2.13.0.

**Comment by jvican** [comment]
GitHub is misbehaving, I think. I'm not getting some comments in this repo and this wasn't closed, even though the fix matches their commit format. Closing.

**Comment by dwijnand** [comment]
The reason this wasn't auto-closed is because the fix is in the 1.0.0 branch, which isn't the default branch of the repo.
